### DISCUSSION TITLE
C++ user module

### DISCUSSION THREAD

**Discussion by ma261065** [micropython/micropython#14259] [ESP32] [answered]
I've been using someone else's code as a c++ user module for a while, following the example at examples/usercmodule/cppexample

However, the maintainer of the original code has recently started using c++ vector

Since the user module declaration has to be compatible with C everything goes in extern "C" scope, and declarations like `vector<uint32_t> mylist` don't work.

Am I correct in understanding that a "C++ User module" can use C++ syntax in the source file, but we are limited to "C compatible" in function declarations etc?

**Comment by bikeNomad** [comment, marked as answer]
You are correct (if I understand you correctly). C++ code can use C++ functions/methods _internally_; it's just that the functions that are called by Python code must be callable from C (no name mangling, etc.). You can ensure this by declaring them as `extern "C"` (with `#ifdef __cplusplus` guards).  You also need to be careful with memory allocation, C++ exceptions and static guards, etc.

You could use a `std::vector` inside your C++ code, but if you wanted to provide it to MicroPython code it'd have to be presented as a `bytes` or `array.array` object or similar. You could use `std::vector::cbegin()` to get a pointer to the start of the allocated memory.

**Reply by cnadler86** [reply]
I have some similar topics. Does anyone have an example?
I want to use a cpp Objekt in my User Module anf don't know how to integrate it...

**Comment by cnadler86** [comment]
I found some example code of a binding for c++ [here](https://github.com/esp-arduino-libs/ESP32_Display_Panel/tree/62223408da2f2c7286ca4c1f203f53110c88f336/mpy_support).

Didn't test, if it works, but I could at least build a module with that structure.

Basically the module registration happens in c, while the User Module is in c++ (locals). For this, the stand Makros `MP_DEFINE_CONST_FUN_OBJ_xxx` were extended in order to be c++ compatible.

Is this a good practice?
